core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hadoop;
+
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.BaseMetastoreCatalog;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+
+/**
+ * HadoopCatalog provides a way to use table names like db.table to work with path-based tables under a common
+ * location. It uses a specified directory under a specified filesystem as the warehouse directory, and organizes
+ * two levels directories that mapped to the database and the table respectively. The HadoopCatalog
+ * takes a URI as the warehouse directory. When creating a table such as $db.$tbl, it creates $db/$tbl under
+ * iceberg/warehouse, and put the table metadata into the directory.
+ *
+ * The HadoopCatalog now supports {@link org.apache.iceberg.catalog.Catalog#createTable},
+ * {@link org.apache.iceberg.catalog.Catalog#dropTable}, the {@link org.apache.iceberg.catalog.Catalog#renameTable}
+ * is not supported yet.
+ *
+ * Note: The HadoopCatalog requires that the underlying file system supports atomic rename.
+ */
+public class HadoopCatalog extends BaseMetastoreCatalog implements Closeable {
+  private static final String ICEBERG_HADOOP_WAREHOUSE_BASE = "iceberg/warehouse";
+  private final Configuration conf;
+  private String warehouseLocation;
+
+  public HadoopCatalog(Configuration conf, String warehouseLocation) {
+    Preconditions.checkArgument(warehouseLocation != null && !warehouseLocation.equals(""),
+        "no location provided for warehouse");
+
+    this.conf = conf;
+    this.warehouseLocation = warehouseLocation.replaceAll("/$", "");
+  }
+
+  public HadoopCatalog(Configuration conf) {
```

Comment:
This constructor should have Javadoc that states how the warehouse path is defaulted.